swim/heal_partition_test.go

```diff
@@ -0,0 +1,475 @@
+		B.ProgressTime(time.Millisecond * 5)
+	}
+
+	hosts := A.Hosts()
+	for _, B := range Bs {
+		hosts = append(hosts, B.Hosts()...)
+	}
+	A[0].node.discoverProvider = statichosts.New(hosts...)
+
+	targets := A[0].node.healer.Heal()
+	assert.Len(t, targets, 5, "expected correct amount of targets")
+
+	waitForConvergence(t, time.Second, A...)
+	for _, B := range Bs {
+		waitForConvergence(t, time.Second, B...)
+	}
+	waitForConvergence(t, time.Second, A...)
+	for _, B := range Bs {
+		waitForConvergence(t, time.Second, B...)
+	}
+
+	targets = A[0].node.healer.Heal()
+	assert.Len(t, targets, 4, "expected correct amount of targets")
+
+	waitForPartitionHeal(t, time.Second, append(Bs, A)...)
+}
+
+// TestPartitionHealWithTime tests whether partitions are healed over time
+// automatically.
+func TestPartitionHealWithTime(t *testing.T) {
+	A := newPartition(t, 5)
+	B := newPartition(t, 5)
+	defer destroyNodes(A...)
+	defer destroyNodes(B...)
+
+	A.AddPartitionWithStatus(B, Faulty)
+	B.AddPartitionWithStatus(A, Faulty)
+
+	A.ProgressTime(time.Millisecond)
+	B.ProgressTime(time.Millisecond)
+
+	A[0].node.discoverProvider = statichosts.New(append(A.Hosts(), B.Hosts()...)...)
+
+	A[0].node.healer.Start()
+
+	// Progress time in a background thread. This will cause the node to
+	// attempt a heal periodically.
+	c := clock.NewMock()
+	A[0].node.clock = c
+	go func() {
+		for {
+			c.Add(time.Second)
+		}
+	}()
+
+	waitForPartitionHeal(t, 3000*time.Millisecond, A, B)
+}
+
+// TestHealBeforeBootstrap tests that we can heal with a node that is still
+// bootstrapping. We put a node in the bootstrapping phase by giving it a
+// host-list that is impossible to join. When it is in this phase, we trigger
+// a heal attempt from another node. The heal should succeed from the other
+// node's perspective.
+func TestHealBeforeBootstrap(t *testing.T) {
+	a := newChannelNode(t)
+	defer a.Destroy()
+
+	b := newChannelNode(t)
+	defer b.Destroy()
+
+	// block a from completing bootstrap
+	block := make(chan struct{})
+	a.node.RegisterListener(on(JoinTriesUpdateEvent{}, func(e Event) {
+		<-block
+	}))
+
+	discoProvider := statichosts.New(a.node.Address(), b.node.Address())
+
+	// bootstrap a and wait for it to be in bootstrap phase
+	go func() {
+		_, _ = a.node.Bootstrap(&BootstrapOptions{
+			DiscoverProvider: discoProvider,
+		})
+	}()
+
+	// bootstrap b then add a to its bootstrap provider
+	bootstrapNodes(t, b)
+	b.node.discoverProvider = discoProvider
+	b.node.disseminator.ClearChanges()
+
+	// check that a is not yet part of b's membership
+	_, has := b.node.memberlist.Member(a.node.Address())
+	assert.False(t, has, "expected that a is not yet part of b's membership")
+
+	// start heal
+	targets := b.node.healer.Heal()
+	assert.Len(t, targets, 0, "expected that b cannot completely heal with a")
+
+	// a is now part of b's membership
+	_, has = b.node.memberlist.Member(a.node.Address())
+	assert.True(t, has, "expected that a is now part of b's membership")
```

Comment:
No, not really. It has to do with `T212923` that is resolved in both ringpop-go and ringpop-node

- ringpop-go: https://github.com/uber/ringpop-go/pull/65
- ringpop-node: https://github.com/uber/ringpop-node/pull/217

The problem here is that the node is being gossiped around as alive while the node its self has not started pinging because it is not able to complete the bootstrapping. This caused non-recovering partitions.

Even though the partition healing would eventually heal the partition I believe it is undesirable to have a node in the ring that is not pinging on its own. So we should prevent nodes that have not bootstrapped completely (eg. blocked in the bootstrapping phase) to join a ring due to partition healing.